logical/framework/openapi.go

```diff
@@ -0,0 +1,548 @@
+	"github.com/mitchellh/mapstructure"
+)
+
+// OpenAPI specification (OAS): https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.2.md
+const OASVersion = "3.0.2"
+
+// NewOASDocument returns an empty OpenAPI document.
+func NewOASDocument() *OASDocument {
+	return &OASDocument{
+		Version: OASVersion,
+		Info: oasInfo{
+			Title:       "HashiCorp Vault API",
+			Description: "HTTP API that gives you full access to Vault. All API routes are prefixed with `/v1/`.",
+			Version:     version.GetVersion().Version,
+			License: oasLicense{
+				Name: "Mozilla Public License 2.0",
+				URL:  "https://www.mozilla.org/en-US/MPL/2.0",
+			},
+		},
+		Paths: make(map[string]*oasPathItem),
+	}
+}
+
+type OASDocument struct {
+	Version string                  `json:"openapi"`
+	Info    oasInfo                 `json:"info"`
+	Paths   map[string]*oasPathItem `json:"paths"`
+}
+
+type oasInfo struct {
+	Title       string     `json:"title"`
+	Description string     `json:"description"`
+	Version     string     `json:"version"`
+	License     oasLicense `json:"license"`
+}
+
+type oasLicense struct {
+	Name string `json:"name"`
+	URL  string `json:"url"`
+}
+
+type oasPathItem struct {
+	Description     string         `json:"description,omitempty"`
+	Parameters      []oasParameter `json:"parameters,omitempty"`
+	Sudo            bool           `json:"x-vault-sudo,omitempty" mapstructure:"x-vault-sudo"`
+	Unauthenticated bool           `json:"x-vault-unauthenticated,omitempty" mapstructure:"x-vault-unauthenticated"`
+	Create          bool           `json:"x-vault-create,omitempty" mapstructure:"x-vault-create"`
+
+	Get    *OASOperation `json:"get,omitempty"`
+	Post   *OASOperation `json:"post,omitempty"`
+	Delete *OASOperation `json:"delete,omitempty"`
+}
+
+// NewOASOperation creates an empty OpenAPI Operations object.
+func NewOASOperation() *OASOperation {
+	return &OASOperation{
+		Responses: make(map[string]*oasResponse),
+	}
+}
+
+type OASOperation struct {
+	Summary     string                  `json:"summary,omitempty"`
+	Description string                  `json:"description,omitempty"`
+	Tags        []string                `json:"tags,omitempty"`
+	Parameters  []oasParameter          `json:"parameters,omitempty"`
+	RequestBody *oasRequestBody         `json:"requestBody,omitempty"`
+	Responses   map[string]*oasResponse `json:"responses"`
+	Deprecated  bool                    `json:"deprecated,omitempty"`
+}
+
+type oasParameter struct {
+	Name        string     `json:"name"`
+	Description string     `json:"description,omitempty"`
+	In          string     `json:"in"`
+	Schema      *oasSchema `json:"schema,omitempty"`
+	Required    bool       `json:"required,omitempty"`
+	Deprecated  bool       `json:"deprecated,omitempty"`
+}
+
+type oasRequestBody struct {
+	Description string      `json:"description,omitempty"`
+	Required    bool        `json:"required,omitempty"`
+	Content     *oasContent `json:"content,omitempty"`
```

Comment:
Is there a way to indicate expected request body formatting? For example, for OIDC support we may end up needing to support urlparams rather than JSON.

Reply:
oasContent is a map with string keys that are description the type. Right now we populate it only with "application/json", but it is completely open to add other types.

Reply:
Given that maps are reference types does it need to be a pointer? (I think `omitempty` will cause json to elide `nil` maps, although I'm not _sure_)

Reply:
`nil` maps will be elided. I think this is legacy from when oasContent was a struct, so I'll depointerize it now.